crossplane-capi-import/README.md

```diff
@@ -0,0 +1,301 @@
+   the compositions for each customer.
+
+   The drawback to this method is that it isn't flexible and relies on a degree
+   of human interaction in the initial stages as we start to develop the library
+
+2. We attempt to hook into crossplanes composition functions. Allow the
+   `nodeGroupName` parameter to accept a list of existing names, then iterate
+   a new resource into the composition based on the values specified in the list.
+
+One drawback that is applicable to either solution is the ability to recognise
+when nodegroups are removed from the cluster. As we're not managing the
+infrastructure, this would have an impact on the cluster resources, potentially
+causing failures both in crossplane and in CAPI as the values in the claim drift
+away from the nodegroups existing as real cloud infrastructure.
+
+This can be mitigated by maintaining that we only need know about the primary
+nodegroup for the cluster, and work with the customer to understand the level
+of reflection that can be maintained for imported clusters.
+
+The reflection of nodegroups is not strictly required for cluster import to be
+successful but is considered a "nice to have" for the customer as it presents
+an opportunity towards a "Single pane of glass" view of their cluster either as
+resources tracked inside the cluster, or through Happa where the resources are
+considered read only.
+
+### CAPI resources
+
+Compositions implementing this RFC will create the following CAPI resources:
+
+- Cluster
+- MachinePool
+
+Additionally cloud specific providers should be created
+
+| AWS                    | Azure                    | GCP                    |
+|------------------------|--------------------------|------------------------|
+| AWSManagedCluster      | AzureManagedCluster      | GCPManagedCluster      |
+| AWSManagedControlPlane | AzureManagedControlPlane | GCPManagedControlPlane |
+| AWSManagedMachinePool  | AzureManagedMachinePool  | GCPManagedMachinePool  |
+
+### `kubeconfig`
+
+The `kubeconfig` secret is normally generated by CAPI controllers and stored
+in the cluster namespace as `<cluster_name>-kubeconfig`
+
+As CAPI should be running in a non-reconcilling mode for imported clusters,
+it is not known if this kubeconfig will be generated automatically by CAPI or
+whether it needs to be imported separately.
+
+To work around potential limitations here, crossplane leverages the
+`cluster-auth` capabilities and stores this secret as `<cluster_name>-kubeconfig-cluster-auth`
+
+This secret contains a slightly different structure from CAPI and places the
+kubeconfig at the secret data key `data.kubeconfig`.
+
+This cannot be controlled via crossplane and in order to make this secret
+compatible with our own deployments via `App-Platform` we may need to enhance
+our `App` CR to accept a secret data key, or include this location in the list
+of key locations checked for a `kubeconfig` value.
```

Comment:
The issue here is that we might depend on the kubeconfig format and naming conventions in other places in our product (e.g. operators like PMO needing to talk to the cluster).

We need to check if that is the case, and might need a plan to "reformat" the secret somehow (I've seen some use cases where people did that via Kyverno or other webhook)

Reply:
I understand the gist of this but have concerns with trying to mutate a secret controlled by crossplane.

As an alternate approach, it would be possible to use either crossplane directly, or ESO to copy the `kubeconfig` value out of the generated secret and map this directly into a CAPI formatted secret.

In this scenario, my preference would be to use crossplane and try and keep the whole process self contained to a single product.